magic guide (#42)
thirdweb-dev/docs: guides/create-magic-wallet/index.mdx

---
authors: [raza]
tags: [magic, wallet]

> date: "2022-05-12"

Author: done

<!-- truncate -->

## Intro

As web3 is growing, more solutions are becoming available to generate and connect wallets. Magic link allows the creation and authentication of a wallet through an email address. In this app, the user can either create or log into the app using just their email address. No setting up accounts, no browser plugins necessary.

Check out our example repository with the project built out on [Github](https://github.com/thirdweb-example/magic-link-next-ts).

## Flow

The user clicks the connect wallet button, enters their email address and receives an email in their inbox. The need to verify themselves by clicking the link inside the email address. Important, the user must not close the tab, while the authentication is going on. Else the user has to start again. When the user clicks on the link in the email, the authentication starts and the app will reload and redirect once the user is authenticated properly.

## Setup

To build this app, we need to walk through the following steps.

1. Build a connector for Magic Link
2. Build the button for the `Connect Wallet` function

> Note prior to starting to build the app, make sure you have a Magic link account setup. Click [here](https://magic.link) to create one.

Reviewer: according to @palerthanale we should not use click "here" links. google finds these spammy.

Instead just link the thing you're referring to correctly. So in this case probably `Magic link account` should be the linked part.

Author: done

Let’s go!

## Building the app

> Go ahead and clone [this](https://github.com/thirdweb-example/next-typescript-starter) repo to get the starters template. We will go through the logic, but won't show every piece of code. If you want to have a look at the full piece of code, check out the [repo](https://github.com/thirdweb-example/magic-link-next-ts)!

Author: done

### Build the connector

> In the `_app.tsx` file we need to include a connection for Magic link. To do this we use the `WalletConnector` provided by thirdweb’s `React` sdk. Note you also need to pass your api key from your Magic link account. 

Reviewer: Keep it less next-js (with typescript) specific imo...

```suggestion
In the root of our application we need to include the options for the Magic Link connector. The only required option for the magic link connector is your magic link `api key` which you will be able to find in your magic link dashboard.
```

Author: done

```tsx
const magicLinkWalletConnector: WalletConnector = {
  name: "magic",
  options: {
    // Replace this with your own magic link api key

>     apiKey: "pk_live_3B2498906537FF2A",

Author: changed it to send that message

    rpcUrls: {
      [ChainId.Mumbai]: "https://mumbai.magic.io/rpc",

>     },

Reviewer: we do not need rpcUrls to be passed into the options here, this is all set up for the user under the hood, the only required option is the apiKey

Reviewer: @jarrodwatts showed me that the react SDK requires this right now, I'll make sure it doesn't going forward

Author: ok

> ```tsx
> const magicLinkWalletConnector: WalletConnector = {
>   name: "magic",
>   options: {
>     // Replace this with your own magic link api key
>     apiKey: "pk_live_3B2498906537FF2A",
>     rpcUrls: {
>       [ChainId.Mumbai]: "https://mumbai.magic.io/rpc",
>     },
>   },
> };

Reviewer: much more digestible if we just adjust it

```suggestion
```jsx
const magicLinkConnectorOptions = {
  name: "magic",
  options: {
    // Replace this with your own magic link api key
    apiKey: "pk_live_3B2498906537FF2A"
  },
};
```

Author: done

```

> Inside your wrapper make sure to include a reference to the connector.

Reviewer: ```suggestion
Make sure to pass the magic link connector option to the `<ThirdwebProvider />` `walletConnectors` prop.
```

... and then show below how we're passing the exact key we defined above into an array with the connectors

Author: done

> ```tsx

Author: done

return (
    <ThirdwebProvider
      desiredChainId={activeChainId}

>       walletConnectors={connectors}

Reviewer: we never set up the `connectors` array here, so this might be confusing?

    >   
      <Component {...pageProps} />
    </ThirdwebProvider>
  );
}
```

> Check out the full file [here](https://github.com/thirdweb-example/magic-link-next-ts/blob/main/pages/_app.tsx).

Reviewer: ```suggestion
Check out the [full file](https://github.com/thirdweb-example/magic-link-next-ts/blob/main/pages/_app.tsx).
```

Author: done

### Create our UI

thirdweb’s sdk has a simple hook to get the connect wallet functionality for Magic link. First we import that and set it up.

> ```tsx
> import { useAddress, useDisconnect, useMagic } from "@thirdweb-dev/react";
> 
> //inside your app
> const Home: NextPage = () => {
> 	const connectWithMagic = useMagic(); // Hook to connect with Magic Link.
> ...

Reviewer: barely any useful types, let's make it jsx instead
```suggestion
```jsx
import { useAddress, useDisconnect, useMagic } from "@thirdweb-dev/react";

//inside your app
const Home = () => {
	const connectWithMagic = useMagic(); // Hook to connect with Magic Link.
...
```

Author: done

```

Next we’ll setup a way for the user to input their email address

> ```tsx
> //inside your app
> const Home: NextPage = () => {
> 	const [email, setEmail] = useState<string>("");

Reviewer: ideally we remove the (very little) types that this has, it's easier to just keep it as javascript. people can follow along more easily:
```suggestion
```jsx
//inside your app
const Home = () => {
	const [email, setEmail] = useState("");
```

Author: ok

```

> ```html

Reviewer: this cannot be `html` because it uses react-props on the `input` component, so `jsx` or `tsx` would be more appropriate here

Author: adjusted

<input
      type="email"
      placeholder="Your Email Address"
      className={styles.textInput}
      onChange={(e) => setEmail(e.target.value)}
      />
```

Finally we’ll pass the email address to our `useMagic` hook to create the wallet and allow the user to start the authentication process.

> ```html

Reviewer: again, can't be `html`

<a className={styles.mainButton} onClick={() => connectWithMagic({ email }) Login </a>
```

## That’s it!

> That’s how you build a functionality that lets the user login with their email address. Again we went over the core logic of the app. Feel free to go to the repo to get a full copy and start using the Magic Link wallet.

Author: done
